atc/db/build_in_memory_check.go

```diff
@@ -0,0 +1,445 @@
+		"build":    b.ID(),
+		"team":     b.TeamName(),
+		"pipeline": b.PipelineName(),
+	}
+
+	if b.resourceId != 0 {
+		data["resource"] = b.resourceName
+	}
+
+	if b.resourceTypeId != 0 {
+		data["resourceType"] = b.resourceTypeName
+	}
+
+	return data
+}
+
+func (b *inMemoryCheckBuild) TracingAttrs() tracing.Attrs {
+	attrs := tracing.Attrs{
+		"build":    fmt.Sprintf("%d", b.ID()),
+		"team":     b.TeamName(),
+		"pipeline": b.PipelineName(),
+	}
+
+	if b.resourceId != 0 {
+		attrs["resource"] = b.resourceName
+	}
+
+	if b.resourceTypeId != 0 {
+		attrs["resourceType"] = b.resourceTypeName
+	}
+
+	return attrs
+}
+
+// Reload just reloads the embedded checkable.
+func (b *inMemoryCheckBuild) Reload() (bool, error) {
+	return b.checkable.Reload()
+}
+
+// AcquireTrackingLock returns a noop lock because in-memory check build runs
+// on only one ATC, thus no need to use a lock to sync among ATCs.
+func (b *inMemoryCheckBuild) AcquireTrackingLock(logger lager.Logger, interval time.Duration) (lock.Lock, bool, error) {
+	return lock.NoopLock{}, true, nil
+}
+
+func (b *inMemoryCheckBuild) Finish(status BuildStatus) error {
+	if !b.running {
+		panic("not a running in-memory-check-build")
+	}
+
+	if !b.reallyRun {
+		return nil
+	}
+
+	tx, err := b.conn.Begin()
+	if err != nil {
+		return err
+	}
+	defer Rollback(tx)
+
+	err = b.saveEvent(tx, event.Status{
+		Status: atc.BuildStatus(status),
+		Time:   time.Now().Unix(),
+	})
+	if err != nil {
+		return err
+	}
+
+	// Release the containers using in this build, so that they can be GC-ed.
+	_, err = psql.Update("containers").
+		Set("in_memory_check_build_id", nil).
+		Where(sq.Eq{"in_memory_check_build_id": b.id}).
+		RunWith(tx).
+		Exec()
+	if err != nil {
+		return err
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		return err
+	}
+
+	err = b.conn.Bus().Notify(buildEventsChannel(b.id))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (b *inMemoryCheckBuild) saveEvent(tx Tx, event atc.Event) error {
+	if !b.running {
+		panic("not a running in-memory-check-build")
+	}
+
+	payload, err := json.Marshal(event)
+	if err != nil {
+		return err
+	}
+
+	_, err = psql.Insert("check_build_events").
+		Columns("event_id", "build_id", "type", "version", "payload").
+		Values(b.eventIdSeq, b.id, string(event.EventType()), string(event.Version()), payload).
+		RunWith(tx).
+		Exec()
+	b.eventIdSeq++
+
+	return err
+}
+
+func (b *inMemoryCheckBuild) Variables(logger lager.Logger, secrets creds.Secrets, varSourcePool creds.VarSourcePool) (vars.Variables, error) {
+	pipeline, found, err := b.Pipeline()
+	if err != nil {
+		return nil, fmt.Errorf("failed to find pipeline: %w", err)
+	}
+	if !found {
+		return nil, errors.New("pipeline not found")
+	}
+
+	return pipeline.Variables(logger, secrets, varSourcePool)
+}
+
+func (b *inMemoryCheckBuild) SaveEvent(ev atc.Event) error {
+	if !b.running {
+		panic("not a running in-memory-check-build")
+	}
+
+	if ev.EventType() == event.EventTypeSelectedWorker {
```

Comment:
This feels pretty hacky since it depends on fairly intimate knowledge of the exec package. I feel like it would be slightly better to at least make this dependency explicit by e.g. adding a `SelectWorker` method to `Build` which `BuildStepDelegate` invokes (rather than `SaveEvent`)

Reply:
Good idea.